flocker/provision/_install.py

```diff
@@ -313,16 +348,85 @@ def task_install_cli(distribution, package_source=PackageSource()):
     return _task_install_commands[distribution](distribution, package_source)
 
 
-def install_cli(package_source, node):
+PIP_CLI_PREREQ_APT = [
+    'gcc',
+    'libffi-dev',
+    'libssl-dev',
+    'python2.7',
+    'python2.7-dev',
+    'python-virtualenv',
+]
+
+PIP_CLI_PREREQ_YUM = [
+    'gcc',
+    'libffi-devel',
+    'openssl-devel',
+    'python',
+    'python-devel',
+    'python-virtualenv',
+]
+
+
+def task_cli_pip_prereqs(package_manager):
     """
-    Return an effect to run the CLI installation tasks on a remote node.
+    Install the pre-requisites for pip installation of the Flocker client.
 
+    :param bytes package_manager: The package manager (apt, dnf, yum).
+    :return: an Effect to install the pre-requisites.
+    """
+    if package_manager in ('dnf', 'yum'):
+        return sudo_from_args(
+            [package_manager, '-y', 'install'] + PIP_CLI_PREREQ_YUM
+        )
+    elif package_manager == 'apt':
+        return sequence([
+            sudo_from_args(['apt-get', 'update']),
+            sudo_from_args(['apt-get', '-y', 'install'] + PIP_CLI_PREREQ_APT),
+        ])
+    else:
+        raise UnsupportedDistribution()
```

Comment:
:question: There's now a couple of different places where the patch executes different code depending on the package manager. It might be worth extracting that out into a couple of classes (one for yum/dnf, one for apt), and making `ensure_minimal_setup` and `task_cli_pip_prereqs` methods on those. Removes the `if`s and associated duplication. 

I really don't think it's a necessary change though.